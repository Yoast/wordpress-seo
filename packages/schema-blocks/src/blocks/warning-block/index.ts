import { WarningBlockAttributes } from "./warningBlockAttributes";

export { WarningBlockAttributes };
