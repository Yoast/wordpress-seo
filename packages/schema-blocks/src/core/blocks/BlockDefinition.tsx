import { ReactElement } from "react";
import { createElement, Fragment } from "@wordpress/element";
import {
	registerBlockType,
	BlockConfiguration,
	BlockEditProps,
	BlockSaveProps,
} from "@wordpress/blocks";
import { InspectorControls, BlockIcon } from "@wordpress/block-editor";
import { select } from "@wordpress/data";
import { PanelBody } from "@wordpress/components";

import BlockInstruction from "./BlockInstruction";
import Definition from "../Definition";
import BlockRootLeaf from "../../leaves/blocks/BlockRootLeaf";
import parse from "../../functions/blocks/parse";
import { registerBlockDefinition } from "./BlockDefinitionRepository";
import logger from "../../functions/logger";
import { openGeneralSidebar } from "../../functions/gutenberg/sidebar";

export interface RenderEditProps extends BlockEditProps<Record<string, unknown>> {
	clientId: string;
	name?: string;
}

export interface RenderSaveProps extends BlockSaveProps<Record<string, unknown>> {
	clientId?: string;
	className?: string;
}

export type MutableBlockConfiguration = {
	-readonly [K in keyof BlockConfiguration]: BlockConfiguration[K]
}

/**
 * BlockDefinition class.
 */
export default class BlockDefinition extends Definition {
	public static separatorCharacters = [ "b", "c", "d", "f", "g", "h", "k", "m", "z" ];
	public static parser = parse;

	public instructions: Record<string, BlockInstruction>;
	public tree: BlockRootLeaf;

	/**
	 * Renders editing the block.
	 *
	 * @param props The props.
	 *
	 * @returns The rendered block.
	 */
	edit( props: RenderEditProps ): ReactElement {
		// Force the sidebar open.
		if ( select( "core/block-editor" ).isBlockSelected( props.clientId ) ) {
			openGeneralSidebar( "edit-post/block", true );
		}

		const sidebarElements = this.sidebarElements( props );

		// Take the children directly to avoid creating too many Fragments.
		const elements = this.tree.children.map( ( leaf, i ) => leaf.edit( props, i ) ).filter( e => e !== null );

		if ( sidebarElements.length > 0 ) {
			const sidebarContainer =
				<InspectorControls key="sidebar-inspector-controls">
					<PanelBody>
						{ sidebarElements }
					</PanelBody>
				</InspectorControls>;

			elements.unshift( sidebarContainer );
		}

		if ( elements.length === 1 ) {
			return elements[ 0 ] as ReactElement;
		}

		return createElement( Fragment, { key: props.clientId }, elements.map( ( element, i ) => createElement( Fragment, { key: i }, element ) ) );
	}

	/**
	 * Renders the persisted block.
	 *
	 * @param props The props.
	 *
	 * @returns The rendered block.
	 */
	save( props: RenderSaveProps ): ReactElement {
		return this.tree.save( props );
	}

	/**
	 * Registers the block with Gutenberg.
	 */
	register(): void {
		const configuration = this.configuration() as MutableBlockConfiguration;
		const name = configuration.name as string;
		delete configuration.name;

		configuration.edit = props => this.edit( props );
		configuration.save = props => this.save( props );

		logger.debug( "registering block " + name );

		if ( configuration.icon && typeof configuration.icon === "string" && configuration.icon.startsWith( "<svg" ) ) {
			configuration.icon = this.createBlockIcon( configuration );
		}

		// Register the block to WordPress.
		registerBlockType( name, configuration );
		// Register the block with our own code.
		registerBlockDefinition( name, this );
	}

	/**
	 * Creates the sidebar elements.
	 *
	 * @param props The properties of the block to create a sidebar for.
	 *
	 * @returns The sidebar element to render.
	 */
	sidebarElements( props: RenderEditProps ): ReactElement[] {
		return Object.values( this.instructions )
			.map( ( instruction, index ) => <Fragment key={ instruction.id }>
				{ instruction.sidebar( props, index ) }
			</Fragment> )
			.filter( e => e !== null );
	}

	/**
	 * Creates a block icon.
	 *
	 * @param configuration The block configuration.
	 *
	 * @returns The sidebar element to render.
	 */
	private createBlockIcon( configuration: MutableBlockConfiguration ): ReactElement {
		const icon = <span
			className="yoast-schema-blocks-icon"
			dangerouslySetInnerHTML={ { __html: configuration.icon as string } }
		/>;

		return <BlockIcon icon={ icon } />;
	}
}
