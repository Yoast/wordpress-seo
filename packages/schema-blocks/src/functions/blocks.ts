import { select } from "@wordpress/data";

/**
 * Gets attributes of inner blocks.
 *
 * @param clientId The client ID of the parent block.
 * @param blocks   A mapping of block name to attribute key.
 *
 * @returns An array contain block names and values.
 */
export function getInnerBlocksAttributes( clientId: string, blocks: Record<string, string> ): { name: string; value: unknown }[] {
	let innerBlocks = select( "core/block-editor" ).getBlock( clientId ).innerBlocks;
	innerBlocks     = innerBlocks.filter( block => block.name in blocks );

	const values = [];

	for ( const block of innerBlocks ) {
		const key   = blocks[ block.name ];
		values.push( { name: block.name, value: block.attributes[ key ] } );
	}

	return values;
}
