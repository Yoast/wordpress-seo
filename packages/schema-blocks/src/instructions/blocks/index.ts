import "./Block";
import "./ClassName";
import "./CurrencySelect";
import "./Date";
import "./Heading";
import "./InheritSidebar";
import "./InnerBlocks";
import "./RichText";
import "./Select";
import "./SidebarCheckbox";
import "./SidebarDuration";
import "./SidebarInput";
import "./SidebarSelect";
import "./TextInput";
import "./Title";
import "./Variation";
import "./VariationPicker";
import "./Duration";
