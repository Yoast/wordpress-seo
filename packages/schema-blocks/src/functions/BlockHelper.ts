import { Block, BlockInstance } from "@wordpress/blocks";
import { dispatch, select } from "@wordpress/data";

/**
 * Gets the inner blocks of the block with the given clientId from the core/block-editor store.
 *
 * @param clientId The clientId of the block you want.
 *
 * @returns {BlockInstance} The block you want.
 */
function getBlockByClientId( clientId: string ): BlockInstance {
	/* istanbul ignore next */
	return select( "core/block-editor" ).getBlock( clientId );
}

/**
 * Removes a block from the editor.
 *
 * @param {string} clientId The client id of the block to remove.
 */
function removeBlock( clientId: string ): void {
	dispatch( "core/block-editor" ).removeBlock( clientId );
}

/**
 * Restores the recommended or required block that had just been removed.
 *
 * @param {string} clientId          The client ID of the warning block.
 * @param {string} removedBlock      The removed block.
 */
function restoreBlock( clientId: string, removedBlock: BlockInstance ): void {
	dispatch( "core/block-editor" ).replaceBlock( clientId, removedBlock );
}

/**
 * Finds the block type in the list of registered blocks in the block editor.
 *
 * @param {string} blockName The block to search for.
 *
 * @return {BlockInstruction} The found block.
 */
function getBlockType( blockName: string ): Block | undefined {
	return select( "core/blocks" ).getBlockType( blockName );
}

/**
 * Retrieves a human readable block name.
 *
 * @param blockName The block name (e.g. the Gutenberg block id).
 *
 * @returns A human readable block title.
 */
function getHumanReadableBlockName( blockName: string ): string {
	const blockType = getBlockType( blockName ) || null;
	if ( blockType ) {
		return blockType.title;
	}

	const lastSlash = blockName.lastIndexOf( "/" );
	if ( lastSlash < 0 || lastSlash === blockName.length - 1 ) {
		return blockName;
	}

	return blockName.substring( lastSlash + 1 ).toLocaleLowerCase();
}

export { getBlockByClientId, removeBlock, restoreBlock, getBlockType, getHumanReadableBlockName };
