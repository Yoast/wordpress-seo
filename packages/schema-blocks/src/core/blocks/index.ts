import BlockDefinition from "./BlockDefinition";
import BlockInstruction from "./BlockInstruction";
import { BlockInstructionClass } from "./BlockInstructionClass";
import BlockLeaf from "./BlockLeaf";
import * as BlockDefinitionRepository from "./BlockDefinitionRepository";

export {
	BlockDefinition,
	BlockInstruction,
	BlockInstructionClass,
	BlockLeaf,
	BlockDefinitionRepository,
};
