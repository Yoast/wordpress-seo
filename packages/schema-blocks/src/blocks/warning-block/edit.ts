import { createElement } from "@wordpress/element";
import { __ } from "@wordpress/i18n";
import { removeBlock, restoreBlock } from "../../functions/BlockHelper";
import { RenderEditProps } from "../../core/blocks/BlockDefinition";
import { BlockInstance } from "@wordpress/blocks";

/**
 * Renders the edit of the warning block.
 *
 * @param {RenderEditProps} props The render edit props of this block.
 *
 * @returns {JSX.Element} The React element.
 */
export function edit( props: RenderEditProps ): JSX.Element {
	const { clientId } = props;

	const { removedBlock, warningText, isRequired } = props.attributes;

	return createElement(
		"div",
		{
			key: "warning-div",
			className: [ "yoast-warning-block", isRequired ? "required" : "recommended" ].join( " " ),
		},
		[
			createElement(
				"p",
				{
					key: "warning-paragraph",
					className: "yoast-warning-block-message",
					dangerouslySetInnerHTML: {
						__html: warningText,
					},
				},
			),
			createElement(
				"div",
				{
					key: "buttons-div",
				},
				[
					createElement(
						"button",
						{
							key: "button-yes",
							onClick: () => {
								removeBlock( clientId );
							},
						},
						__( "Yes", "yoast-schema-blocks" ),
					),
					createElement(
						"button",
						{
							key: "button-no",
							onClick: () => {
								restoreBlock( clientId, removedBlock as BlockInstance );
							},
						},
						__( "No, please undo this", "yoast-schema-blocks" ),
					),
				],
			),
		],
	);
}
