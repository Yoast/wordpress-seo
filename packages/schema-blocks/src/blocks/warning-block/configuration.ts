import { BlockConfiguration } from "@wordpress/blocks";
import { edit } from "./edit";

export const WarningBlock: BlockConfiguration = {
	/**
	 * The title. It's not translatable, since we don't want to show the warning block in the block inserter.
	 */
	title: "Warning",

	/**
	 * The category is there to help users browse and discover blocks.
	 * We actually don't want the warning block to be discoverable, but category is not optional.
	 */
	category: "common",

	/**
	 * The attributes.
	 */
	attributes: {
		removedBlock: {
			type: "object",
		},
		warningText: {
			type: "string",
		},
		isRequired: {
			type: "boolean",
		},
	},

	/**
	 * Make sure that users cannot add this block manually.
	 */
	supports: {
		inserter: false,
	},

	/**
	 * Renders editing the warning block.
	 *
	 * @param props The properties.
	 *
	 * @return The rendered component.
	 */
	edit,

	/**
	 * Renders null, because the warning block isn't output on the frontend.
	 *
	 * @return null
	 */
	save(): JSX.Element {
		return null;
	},
};
