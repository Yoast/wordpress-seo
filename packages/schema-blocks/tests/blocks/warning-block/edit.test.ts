import * as renderer from "react-test-renderer";
import { shallow } from "enzyme";

import { removeBlock, restoreBlock } from "../../../src/functions/BlockHelper";
import { edit } from "../../../src/blocks/warning-block/edit";
import { RenderEditProps } from "../../../src/core/blocks/BlockDefinition";

jest.mock( "../../../src/functions/BlockHelper", () => {
	return {
		removeBlock: jest.fn(),
		restoreBlock: jest.fn(),
	};
} );

describe( "The edit function", () => {

	const defaultProps: RenderEditProps = {
		clientId: "1234-abcd",
		className: "some-class",
		isSelected: false,
		setAttributes: jest.fn(),
		attributes: {
			removedBlock: "yoast/recipe",
			removedAttributes: {
				className: "yoast-recipe",
			},
			warningText: "Are you sure you want to remove this block?",
			isRequired: true,
		},
	};

	it( "should render the warning block.", () => {
		const tree = renderer
			.create( edit( defaultProps ) )
			.toJSON();

		expect( tree ).toMatchSnapshot();
	} );

	it( "should render the warning block when the removed block was not required.", () => {
		const props = Object.assign( {}, defaultProps );
		// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
		// @ts-ignore -- We overwrite the attributes for testing purposes.
		props.attributes.isRequired = false;

		const tree = renderer
			.create( edit( props ) )
			.toJSON();

		expect( tree ).toMatchSnapshot();
	} );

	it( "should call the `removeBlock` function if the 'yes' button is clicked.", () => {
		const tree = shallow( edit( defaultProps ) );

		const yesButton = tree.find( "button" ).first();

		yesButton.simulate( "click" );

		expect( removeBlock ).toHaveBeenCalled();
	} );

	it( "should call the `restoreBlock` function if the 'no' button is clicked.", () => {
		const tree = shallow( edit( defaultProps ) );

		const noButton = tree.find( "button" ).at( 1 );

		noButton.simulate( "click" );

		expect( restoreBlock ).toHaveBeenCalled();
	} );
} );
