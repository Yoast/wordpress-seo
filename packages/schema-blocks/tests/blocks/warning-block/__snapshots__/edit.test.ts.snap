// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The edit function should render the warning block when the removed block was not required. 1`] = `
<div
  className="yoast-warning-block recommended"
>
  <p
    className="yoast-warning-block-message"
    dangerouslySetInnerHTML={
      Object {
        "__html": "Are you sure you want to remove this block?",
      }
    }
  />
  <div>
    <button
      onClick={[Function]}
    >
      Yes
    </button>
    <button
      onClick={[Function]}
    >
      No, please undo this
    </button>
  </div>
</div>
`;

exports[`The edit function should render the warning block. 1`] = `
<div
  className="yoast-warning-block required"
>
  <p
    className="yoast-warning-block-message"
    dangerouslySetInnerHTML={
      Object {
        "__html": "Are you sure you want to remove this block?",
      }
    }
  />
  <div>
    <button
      onClick={[Function]}
    >
      Yes
    </button>
    <button
      onClick={[Function]}
    >
      No, please undo this
    </button>
  </div>
</div>
`;
